### PR TITLE
Adding 'pluginRuntimeVersion' to PluginDescriptor

### LINKED ISSUE #3547: Include the plugin library version with plugin descriptor to determine library version of the plugin
**Describe the feature request**
As a tanzu cli core developer, I would like to know the version of the plugin library(tanzu-plugin-runtime) version used to develop a plugin.

Provide a means so that tanzu cli can determine the plugin library version that was used to develop a particular plugin.

Currently, with `tanzu <plugin-name> info` command, we are listing plugin descriptor information.
```
~ $ tanzu cluster info
{"name":"cluster","description":"Kubernetes cluster operations","version":"v0.27.0-dev","buildSHA":"e28420e75","digest":"","group":"Run","docURL":"","completionType":0,"aliases":["cl","clusters"],"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":""}
```
We should also include the plugin library version along with this output so that tanzu cli can determine the library version and use that for future support validation.

**Describe alternative(s) you've considered**

**Affected product area (please put an X in all that apply)**

- ( ) APIs
- ( ) Addons
- (x) CLI
- ( ) Docs
- ( ) IAM
- ( ) Installation
- (x) Plugin
- ( ) Security
- ( ) Test and Release
- ( ) User Experience
- ( ) Developer Experience

**Additional context**

### PR DESCRIPTION
**Pull request by prkalle**
Signed-off-by: Prem Kumar Kalle <pkalle@vmware.com>

### What this PR does / why we need it
This PR adds "PluginRuntimeVersion" to PluginDescriptor. This version would be used by tanzu core CLI to add version validations.

### Which issue(s) this PR fixes

Fixes #3547 

### Describe testing done for PR

```
❯ tanzu management-cluster info
{"name":"management-cluster","description":"Kubernetes management cluster operations","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Run","docURL":"","completionType":0,"aliases":["mc","management-clusters"],"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu feature  info
{"name":"feature","description":"Operate on features and featuregates","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Run","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu kr info
{"name":"kubernetes-release","description":"Kubernetes release operations","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Run","docURL":"","completionType":0,"aliases":["kr","kubernetes-releases"],"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu login info
{"name":"login","description":"Login to the platform","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"System","docURL":"","completionType":0,"aliases":["lo","logins"],"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu package info
{"name":"package","description":"Tanzu package management","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Run","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu pinniped-auth info
{"name":"pinniped-auth","description":"Pinniped authentication operations (usually not directly invoked)","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Run","docURL":"","hidden":true,"completionType":0,"aliases":["pa","pinniped-auths"],"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu secret info
{"name":"secret","description":"Tanzu secret management","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Run","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

 ❯ tanzu telemetry info
{"name":"telemetry","description":"configure cluster-wide settings for vmware tanzu telemetry","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Run","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu builder info
{"name":"builder","description":"Build Tanzu components","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Admin","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu codegen info
{"name":"codegen","description":"Tanzu code generation tool","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Admin","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

❯ tanzu test info
{"name":"test","description":"Test the CLI","version":"v0.27.0-dev","buildSHA":"74a09a79-dirty","digest":"","group":"Admin","docURL":"","completionType":0,"installationPath":"","discovery":"","scope":"","status":"","discoveredRecommendedVersion":"","pluginRuntimeVersion":"v0.27.0-dev"}

```

### Release note

```release-note
Include the plugin library version with plugin descriptor to determine library version of the plugin

```

### CODE REVIEW

**Comment by anujc25** [comment]
nit: Update filename from `cli/runtime/version/zz_bundled_plugin_runtime_version.go` to `cli/runtime/version/zz_generated_plugin_runtime_version.go`?

**Review by marckhouzam — COMMENTED** [review]
Thanks @prkalle.  That solution works nicely.

Since I wasn't familiar with this approach and before seeing your new PR, I was trying to think of how we could figure out the version of the runtime lib that what used when a plugin was compiled.  So I started wondering if Go could tell us.  I found the `ReadBuildInfo()` function of the `runtime/debug` Go package: https://stackoverflow.com/a/59276408

I think using this package would allow us to avoid any special build steps and simply read the dependency that was used by the plugin directly in the `info` command.

I think another nice advantage is if a plugin uses a specific commit, or the main branch directly, we would get the specific commit in the version, instead of a generic "vx.y.x-dev"

If we prefer the current solution: 
1. What will be the workflow to set a new version of the cli-runtime lib?
1. It may require changing `.github/workflows/dev_tag_retrieval_pr.yaml` for the automatic PR to include modifying the new `zz_bundled_plugin_runtime_version.go` file.
1. Maybe also `hack/verify-dirty.sh` although I am not sure.

File: cli/runtime/apis/cli/v1alpha1/catalog_types.go [hunk 1]
```diff
@@ -156,6 +156,9 @@ type PluginDescriptor struct {
 
 	// PostInstallHook is function to be run post install of a plugin.
 	PostInstallHook Hook `json:"-" yaml:"-"`
+
+	// PluginRuntimeVersion of the plugin. Must be a valid semantic version https://semver.org/
+	PluginRuntimeVersion string `json:"pluginRuntimeVersion" yaml:"pluginRuntimeVersion"`
```
[the remark this answers is not in the document]
Comment (prkalle) [reply cli/runtime/apis/cli/v1alpha1/catalog_types.go:161]: The reason for adding this field in the PluginDescriptor is beacuse tannzu cli can get the properties of a plugin while invoking the plugin and can impose the compatibility conditions (ex: cli vX.Y.Z doesn't support the plugins created with plugin runtime version vA.B.C.

**Comment by prkalle** [comment]
> I found the `ReadBuildInfo()` function of the `runtime/debug` Go package: https://stackoverflow.com/a/59276408
Thank you @marckhouzam for the information(I was not aware of it before). It is good solution and I think we can use this approach once the cli/runtime moves to different repository because, as we are using a `replace` in go.mod of plugins(eg: cluster/management-cluster etc..) the plugin verison is showing as `"v0.0.0-00010101000000-000000000000"`.  

since we are going with the current approach , I did fixed `.github/workflows/dev_tag_retrieval_pr.yam` and `hack/verify-dirty.sh` files as mentioned in your comments.

File: cli/runtime/version/zz_generated_plugin_runtime_version.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+package version
+
+// IMPORTANT: Please !!DO NOT!!! change the content of this file manually
+// This file is auto-generated based on the version details provided during build time
+// and the values mentioned here are used to store the library version
+
+// All the variables are set during build time
+// Note: Please !!DO NOT!!! change the name or remove this variable
+var (
+	Version string = "v0.27.0-dev"
```
Comment (marckhouzam) [inline cli/runtime/version/zz_generated_plugin_runtime_version.go:13]: this should now be "v0.28.0-dev" I think

Comment (anujc25) [reply cli/runtime/version/zz_generated_plugin_runtime_version.go:13]: Is there a way to catch this thing as part of CI? I see we are ignoring this file as part of `verify-dirty.sh`, if we do not skip this file, will it create any issue?

**Comment by anujc25** [comment]
Closing this as requested changes are done as part of another PR: #3641

**Pull request closed without merging**